# Genome analysis of *Pterois miles* - UNDER CONSTRUCTION
This repository contains scripts and implementations that have been developed during my master thesis on the genome analysis of devil firefish *Pterois miles*. Analysis consists of *de novo* construction of genome assembly, genome annotation (transposable elements, gene prediction, functional), phylogenomic tree inferrence and comparative genomic studies. Scripts, workflows and documentation files for each step are provided in their corresponding folders.

All pieces of work were conducted at the [Genomics & Bioinformatics Group](https://genomenerds.her.hcmr.gr/) in IMBBC, of Hellenic Centre for Marine Research.

1. [Construction of genome assembly](https://github.com/genomenerds/SnakeCube)
3. Genome annotation
    - [Transposable elements](https://github.com/ckitsoulis/Pterois-miles-Genome/tree/main/TE_annotation)
    - [Structural (gene prediction)](https://github.com/ckitsoulis/Pterois-miles-Genome/tree/main/Structural_annotation)
    - Functional
5. Comparative genomics


## 
Connected repositories:

[Genomics & Bioinformatics Group](https://github.com/genomenerds)

[SnakeCube](https://github.com/genomenerds/SnakeCube)

[ELDAR](https://github.com/ckitsoulis/ELDAR)


## Acknowledgements
This research was supported through computational resources provided by IMBBC (Institute of Marine Biology, Biotechnology and Aquaculture) of the HCMR (Hellenic Centre for Marine Research). Funding for establishing the [IMBBC HPC](https://hpc.hcmr.gr/) has been received by the MARBIGEN (EU Regpot) project, LifeWatchGreece RI, and the CMBR (Centre for the study and sustainable exploitation of Marine Biological Resources) RI.

## References
1. Flynn J.M., Hubley R., Goubert C., Rosen J., Clark A.G., Feschotte C., Smit A.F. (2020). RepeatModeler2 for automated genomic discovery of transposable element families. Proc Natl Acad Sci U S A. 117(17): 9451-9457. doi: 10.1073/pnas.1921046117.
2. Ou S., Su W., Liao Y., Chougule K., Agda J.R.A., Hellinga A.J., Lugo C.S.B., Elliott T.A., Ware D., Peterson T., Jiang N., Hirsch C.N. and Hufford M.B. (2019). Benchmarking Transposable Element Annotation Methods for Creation of a Streamlined, Comprehensive Pipeline. Genome Biol. 20(1): 275.
3. Haidong Yan, Aureliano Bombarely, Song Li 2020 DeepTE: a computational method for de novo classification of transposons with convolutional neural network. Bioinformatics, Volume 36, Issue 15, 1 August 2020, Pages 4269–4275.
4. Tarailo-Graovac M., Chen N. (2009). Using RepeatMasker to identify repetitive elements in genomic sequences. Curr Protoc Bioinformatics.Chapter 4, Unit 4.10. doi: 10.1002/0471250953.bi0410s25
5. Wu T.D. and Watanabe C.K. (2005). Gmap: a genomic mapping and alignment program for mrna and est sequences. Bioinformatics, 21(9):1859–1875. ISSN 1367-4803. doi:10.1093/bioinformatics/bti310.
6. Buchfink B., Xie C., and Huson D.H. (2015). Fast and sensitive protein alignment using diamond. Nature Methods, 12(1):59–60. ISSN 1548-7105. doi: 10.1038/nmeth.3176.
7. Venturini L., Caim S., Kaithakottil G.G., Mapleson D.L., and Swarbreck D. (2018). Leveraging multiple transcriptome assembly methods for improved gene structure annotation. GigaScience, 7(8). ISSN 2047-217X. doi: 10.1093/gigascience/giy093.
8. Stanke M., Diekhans M., Baertsch R., and Haussler D. (2008). Using native and syntenically mapped cdna alignments to improve de novo gene finding. Bioinformatics, 24(5):637–64. ISSN 1367-4803. doi: 10.1093/bioinformatics/btn013.
9. Haas D.J., Delcher A.L., Mount S.M., Wortman J.R., Smith, Jr R.K., Hannick L.I., Maiti R., Ronning C.M., Rusch D.B., Town C.D., Salzberg S.L., and White O. (2003). Improving the arabidopsis genome annotation using maximal transcript alignment assemblies. Nucleic Acids Res., 31(19):5654–5666. doi: 10.1093/nar/gkg770.
10. Manni M., Berkeley M.R., Seppey M., Simão F.A., and Zdobnov E.M. (2021) BUSCO Update: Novel and Streamlined Workflows along with Broader and Deeper Phylogenetic Coverage for Scoring of Eukaryotic, Prokaryotic, and Viral Genomes. Molecular Biology and Evolution, 38(10):4647–4654. ISSN 1537-1719. doi: 10.1093/molbev/msab199.
11. Quinlan A.R. and Hall I.M.. (2010). Bedtools: a flexible suite of utilities for comparing genomic features. Bioinformatics, 26(6):841–842. ISSN 1367-4803. doi: 10.1093/bioinformatics/btq033.
12. https://github.com/TransDecoder/TransDecoder
13. https://github.com/nathanweeks/exonerate
14. https://github.com/NBISweden/AGAT
15. https://github.com/gpertea/gffread
